diff -urN dbus-1.6.12.old/bus/session.conf.in dbus-1.6.12/bus/session.conf.in
--- dbus-1.6.12.old/bus/session.conf.in	2013-02-12 11:45:32.000000000 +0000
+++ dbus-1.6.12/bus/session.conf.in	2015-02-10 19:28:00.203226659 +0000
@@ -52,8 +52,8 @@
   <limit name="max_message_unix_fds">4096</limit>
   <limit name="service_start_timeout">120000</limit>  
   <limit name="auth_timeout">240000</limit>
-  <limit name="max_completed_connections">100000</limit>  
-  <limit name="max_incomplete_connections">10000</limit>
+  <limit name="max_completed_connections">900</limit>
+  <limit name="max_incomplete_connections">92</limit>
   <limit name="max_connections_per_user">100000</limit>
   <limit name="max_pending_service_starts">10000</limit>
   <limit name="max_names_per_connection">50000</limit>
